examples/tasks/advanced/run.py

```diff
@@ -1,35 +1,49 @@
 """
-Simple example of a task application creating tasks and panes from traits 
-components. 
+Simple example of an advanced task application creating tasks and panes from
+traits components. Compared to the basic example, this version creates a real
+TaskApplication to manage the GUI part. Additionally, the code editor uses a
+tab viewer to support opening multiple files at once.
 
-Note: Run it with 
+Note: Run it with
 $ ETS_TOOLKIT='qt4' python run.py
 as the wx backend is not supported yet for the TaskWindow.
 """
-# Enthought library imports.
-from pyface.api import GUI
-from pyface.tasks.api import TaskWindow
+from pyface.tasks.api import TaskApplication
 
-# Local imports.
 from example_task import ExampleTask
 
 
+def create_task_window(application, files=()):
+    """ Create a new task and open a window for it. """
+    task = ExampleTask()
+    window = application.create_task_window(task)
+
+    # open files passed in as arguments
+    for filename in files:
+        try:
+            task._open_file(filename)
+        except Exception as exc:
+            window.error(
+                message="Can't open '{}'".format(filename),
+            )
+
+
 def main(argv):
-    """ A simple example of using Tasks.
-    """
-    # Create the GUI (this does NOT start the GUI event loop).
-    gui = GUI()
+    """ A more advanced example of using Tasks. """
+    app = TaskApplication(
+        id="MyTaskApplication",
+        name="Python Editor",
+        window_size=(800, 600),
+    )
 
-    # Create a Task and add it to a TaskWindow.
-    task = ExampleTask()
-    window = TaskWindow(size=(800, 600))
-    window.add_task(task)
+    # hook up listener to application initialized event
+    def app_started(event):
+        create_task_window(event.application, files=argv[1:])
 
-    # Show the window.
-    window.open()
+    app.on_trait_change(app_started, "application_initialized")
```

Comment:
The only thing I am still not happy with in this PR is the complexity of this main, with 2 helper functions and the need for this on_trait_change @corranwebster . In my opinion, this API isn't reasonable because it requires to know so many things:
1. you have to listen to application initialized,
2. you have to know that the event has an attribute that gives you back the application object, 
3. dynamic listeners like this isn't the most commonly known feature of traits (for example not something we/Enthought teach in classes)

The way I implemented the window creation in my version of `TaskApplication`, was by overriding the start method, call super and launch window creation afterwards. It is working well enough, though looking at the chain of events precisely, this approach listening to `application_initialized` may be cleaner. If that's the case, I would much rather add complexity in `TaskApplication` than in every subclass. So I would propose to move the listener to `TaskApplication`, calling back a method that does nothing, and that is supposed to be overridden.

The resulting user code could be something like:
```
class MyTaskApplication(TaskApplication):
    def on_application_initialized(self):
        task = ExampleTask()
        self.create_task_window(task)
        for filename in self.initial_files:
            task._open_file(filename)

def main():
    initial_files = sys.argv[1:]
    app = MyTaskApplication(initial_files=initial_files)
    app.run()
```
Do you like that?

Reply:
The reason for wanting to emphasize the event listener code is because this is the preferred way to handle "do this when everything is set up and the event loop is running" from an Envisage plugin's `start` method to avoid the particular hell that is getting the plugin order exactly right.

Unfortunately in the past couple of weeks I've been playing with things, and although the callback method works very nicely on Qt, it fails abysmally under Wx because Wx _requires_ that there is at least one window before the `wx.App` starts, and will quit automatically when the last window closes.

This means that any Wx-based app _must_ create a window as part of its `start` method.

So this means something of a re-design which will probably mean adding a `_create_windows` method that is called before the mainloop starts and which:

* is intended to be overridden in the`Application` class
* has a default implementation for `TaskApplication` which creates a task window (or more enthusiastically, follows the logic for task layouts from the Envisage version).

We can revisit the discussion after I've put something together.

Reply:
Thanks, lots of things I didn't know here. Not sure I understand your focus on what envisage expects though (first paragraph) since this Application object is only useful when people don't use Envisage. I have never seen any issues using this class and creating windows in the `start` method under `qt` (`pyside` to be precise). Are you expecting this to fail sometimes? Otherwise, that could be the way to do things which would work for both Qt and WX.